### PR TITLE
CORE-10386 Re-execute named query until result set is filled

### PR DESCRIPTION
**Pull request by nkovacsx**
### Overview

Change the main named query executor logic to the following:
1. Execute the query for the first time and filter the results.  If the `limit`(page size) is filled up, return the results.
2. If not, then execute the query once again with the original `limit` and repeat until either
- the page size is reached after filtering 
- we run out of records from the database
- we reach the maximum number of retries (currently 5)

### Test plan

Tested using the e2e test cases from https://github.com/corda/corda-e2e-tests/pull/195 and verified the logs to trace that the logic works as expected.

### CODE REVIEW

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 1]
```diff
@@ -31,6 +31,9 @@ class VaultNamedQueryExecutorImpl(
         const val UTXO_TX_COMPONENT_TABLE = "utxo_transaction_component"
         const val TIMESTAMP_LIMIT_PARAM_NAME = "Corda_TimestampLimit"
 
+        // TODO Should this be configurable or hard-coded?
```
Comment (relyafi) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: Should take a decision on this now, and raise (and reference) a ticket if we do want to be configurable but aren't doing yet. I'd be tempted to leave it fixed for now.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 2]
```diff
@@ -75,17 +75,73 @@ class VaultNamedQueryExecutorImpl(
         // Return the filtered/transformed/collected (if present) result to the caller
         return EntityResponse.newBuilder()
             .setResults(collectedResults.map { ByteBuffer.wrap(serializationService.serialize(it).bytes) })
-            .setMetadata(KeyValuePairList(listOf(KeyValuePair("numberOfRowsFromQuery", contractStateResults.size.toString()))))
+            // TODO Should `numberOfRowsFromQuery` be the number of rows before or after filtering?
```
Comment (relyafi) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: What's the intended use of this metadata? If it's for the benefit of the caller, I'd say after (the in-memory filtering is a black box, anything on the API should reflect the end result, not the internal processing).

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 3]
```diff
@@ -75,17 +75,73 @@ class VaultNamedQueryExecutorImpl(
...
+            // If we can't fetch more states we just return the result set as-is
+            if (contractStateResults.isEmpty())
+                break
+
+            // Increase the internal offset, so we don't fetch the same results multiple times
+            internalOffset += contractStateResults.size
+
+            // Apply filters (if there's any) and add the filtered results to the final result set
+            filteredResults.addAll(contractStateResults.filter {
+                vaultNamedQuery.filter?.filter(it, deserializedParams) ?: true
```
Comment (relyafi) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: How common do we expect filters to be? If no filter is a relatively common case, might want to do the check first and skip the filter step here if so (means doing the check once, rather than for each element returned).

Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: Good point actually!

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 4]
```diff
@@ -126,12 +124,15 @@ class VaultNamedQueryExecutorImpl(
...
 
             // Apply filters (if there's any) and add the filtered results to the final result set
-            filteredResults.addAll(contractStateResults.filter {
-                vaultNamedQuery.filter?.filter(it, deserializedParams) ?: true
-            })
+            vaultNamedQuery.filter?.let { vaultNamedQueryFilter ->
+                filteredResults.addAll(contractStateResults.filter {
+                    vaultNamedQueryFilter.filter(it, deserializedParams)
+                })
+            }
```
Comment (relyafi) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: Doesn't this now return no results if we don't have a filter?

**Review by relyafi — APPROVED** [review]

**Review by relyafi — DISMISSED** [review]

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 5]
```diff
@@ -31,6 +31,8 @@ class VaultNamedQueryExecutorImpl(
         const val UTXO_TX_COMPONENT_TABLE = "utxo_transaction_component"
         const val TIMESTAMP_LIMIT_PARAM_NAME = "Corda_TimestampLimit"
 
+        const val RESULT_SET_FILL_RETRY_LIMIT = 5
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:34]: Maybe `RESULT_SET_REPEAT_QUERY_LIMIT` might be better?

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 6]
```diff
@@ -75,17 +74,77 @@ class VaultNamedQueryExecutorImpl(
...
+     *
+     * This logic will be repeated until either:
+     * - The size of the filtered results reaches the "page size"
+     * - We run out of data to fetch from the database
+     * - We reach the number of retries ([RESULT_SET_FILL_RETRY_LIMIT])
+     *
+     * If any of these conditions happen, we just return the result set as-is without filling
+     * up the "page size".
+     */
+    private fun filterResultsAndFillPageSize(
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:103]: Thanks for adding this comment, I was going to add a comment asking you to do this (which I then deleted 😄).

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt (hunk 5 above)
Comment (lankydan) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:34]: The `FILL` part is quite good though. Maybe `RESULT_SET_FILL_PAGE_REQUERY_LIMIT` gives all the context we'd need?

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 7]
```diff
@@ -75,17 +74,77 @@ class VaultNamedQueryExecutorImpl(
         // Return the filtered/transformed/collected (if present) result to the caller
         return EntityResponse.newBuilder()
             .setResults(collectedResults.map { ByteBuffer.wrap(serializationService.serialize(it).bytes) })
-            .setMetadata(KeyValuePairList(listOf(KeyValuePair("numberOfRowsFromQuery", contractStateResults.size.toString()))))
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:78]: The code a few lines up:

```kotlin
val collectedResults = vaultNamedQuery.collector?.collect(
            filteredAndTransformedResults,
            deserializedParams
        )?.results?.filterNotNull() ?: filteredAndTransformedResults
```

Originally I was thinknig that we could also decide to stop querying if the results of running a `VaultNamedQueryCollector` returned `isDone == true`, but the more I think about it, I feel that the collector doesn't really fit correctly. Mainly that the `resultSet` it takes in would need to contain the rows of previous query executions to really "collect" things (while still allowing the query to be re-executed).

I'm ok with not refactoring the collection part into the requerying logic you've added. We can always come back to this in the future if we decide to add it, or we might possibly deprecate the collector interface since I'm not really happy with it when I reflect on it.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 8]
```diff
@@ -75,17 +74,77 @@ class VaultNamedQueryExecutorImpl(
...
+                offsetOverride = internalOffset
+            )
+
+            // If we have no filter, there's no need to continue the loop
+            if (vaultNamedQuery.filter == null)
+                return contractStateResults
+
+            // If we can't fetch more states we just return the result set as-is
+            if (contractStateResults.isEmpty())
+                break
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:135]: I'm not _that_ keen on the `break`, but I'm not sure if any other way of writing it would be any better.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 9]
```diff
@@ -75,17 +74,77 @@ class VaultNamedQueryExecutorImpl(
...
+    ): List<StateAndRef<ContractState>> {
+        val filteredResults = mutableListOf<StateAndRef<ContractState>>()
+
+        var internalOffset = request.offset
+        var currentRetry = 0
+
+        while (filteredResults.size < request.limit && currentRetry < RESULT_SET_FILL_RETRY_LIMIT) {
+            ++currentRetry
+
+            log.debug("Executing try: $currentRetry, fetched ${filteredResults.size} number of results so far.")
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: Can we move this down to `trace` for now.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 10]
```diff
@@ -75,17 +74,77 @@ class VaultNamedQueryExecutorImpl(
...
+
     /**
      * A function that fetches the contract states that belong to the given transaction IDs. The data stored in the
      * component table will be deserialized into contract states using component groups.
      */
     private fun fetchStateAndRefs(
         request: FindWithNamedQuery,
-        whereJson: String?
+        whereJson: String?,
+        offsetOverride: Int? = null
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: I think change `offsetOverride` to `offset` and always pass it into `fetchStateAndRefs`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 11]
```diff
@@ -75,17 +74,77 @@ class VaultNamedQueryExecutorImpl(
...
-            .setMetadata(KeyValuePairList(listOf(KeyValuePair("numberOfRowsFromQuery", contractStateResults.size.toString()))))
+            .setMetadata(KeyValuePairList(listOf(KeyValuePair("numberOfRowsFromQuery", filteredAndTransformedResults.size.toString()))))
             .build()
     }
 
+    /**
+     * This function will fetch a given number ("page size") of records from the database
+     * (amount defined by [request]'s limit field).
+     *
+     * After fetching those records the in-memory filter will be applied. If the filtering
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:88]: I think we're missing some data being passed back to the flow here.

If we're changing the offset under the hood, then we need to communicate this back to the flow. If we don't do this then when we query for the next page (from the flow side) then we're going to re-query data that we've already been through on the previous iteration.

This will relate to the changes that @relyafi has or is going to propose around querying (don't know where things are atm). We'll either need to pass an offset to the flow side (I introduced a metadata map in the avro response to make adding these changes easier), or update the timestamps we're using (but I don't think we've switched to that yet).

It _might_ be possible to the `numberOfRowsFromQuery` that I originally added to infer an offset, but that will need confirming.

I don't think you hit this scenario in your tests as it looks like you hit the page size eventually from one call to `query.execute`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt (hunk 10 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: Yeah I thought about this too thanks!

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt (hunk 8 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:135]: Yeah same, this is more of a "Java-ish" code with the break/return in the while loop but even though I spent some time thinking about this couldn't really come up with a better solution.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/external/events/FindAllExternalEventFactory.kt [hunk 12]
```diff
@@ -33,9 +33,12 @@ class FindAllExternalEventFactory: ExternalEventFactory<FindAllParameters, Entit
     }
 
     override fun resumeWith(checkpoint: FlowCheckpoint, response: EntityResponse): ResultSetExecutor.Results {
+        val numberOfRowsFromQuery = response.metadata.items.single { it.key == "numberOfRowsFromQuery" }.value.toInt()
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/external/events/FindAllExternalEventFactory.kt:36]: I think we'lll need a comment somewhere about why this works.

Let me write down what I think happens based on this and @nkovacsx can you confirm that is what you wanted?

- We do the querying and repeat querying on the database worker side.
- Each time we re-query (to fill the limit/page depending on in-memory filtering) we increase `numberOfRowsFromQuery`. So if we re-query x2 and the limit is 5 then we'll get a value for `numberOfRowsFromQuery` between 5-10. 10 == there could be more, <10 == there isn't another page.
- We then update the `hasNext` and `newOffset` based on this information.
- We can use `numberOfRowsFromQuery` as the `newOffset` because:
  - If we got less than the limit then we stop querying.
  - If we got equal to the limit then we updated the new offset by an amount == to the limit.
  - If we re-queried to fill the page, then we'll get a value higher than the limit, so setting `numberOfRowsFromQuery` as the offset makes sense because we want to jump to the last row we actually queried.

I don't think `hasNext = limit in 1..numberOfRowsFromQuery` is completely right. If we re-query to fill the page and reach the end of the actual data in the database, then we'll request another page from the flow side because we don't know that we reached the end. So say the limit is 5, we re-query to fill the page after filtering, which gives us a total of 9 rows returned in total (before filtering). `hasNext = 5 in 1..9` is true. But we know that there is no more data.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/external/events/FindAllExternalEventFactory.kt [hunk 13]
```diff
@@ -33,9 +33,11 @@ class FindAllExternalEventFactory: ExternalEventFactory<FindAllParameters, Entit
...
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/external/events/FindAllExternalEventFactory.kt:36]: We can probably revert the changes to this file based on the recent changes.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/external/events/NamedQueryExternalEventFactory.kt [hunk 14]
```diff
@@ -34,9 +34,11 @@ class NamedQueryExternalEventFactory : ExternalEventFactory<NamedQueryParameters
     }
 
     override fun resumeWith(checkpoint: FlowCheckpoint, response: EntityResponse): ResultSetExecutor.Results {
+        val numberOfRowsFromQuery = response.metadata.items.single { it.key == "numberOfRowsFromQuery" }.value.toInt()
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/external/events/NamedQueryExternalEventFactory.kt:37]: Can probably revert.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt [hunk 15]
```diff
@@ -15,6 +15,16 @@ data class ResultSetImpl<R> internal constructor(
     private val resultSetExecutor: ResultSetExecutor<R>
 ) : PagedQuery.ResultSet<R> {
 
+    init {
+        // Ideally this will never happen, but we keep this check in here for safety
+        require(offset >= 0) {
+            "Offset cannot be negative or zero"
+        }
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt]: Check doesn't match the message.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt [hunk 16]
```diff
@@ -32,8 +42,20 @@ data class ResultSetImpl<R> internal constructor(
             throw NoSuchElementException("The result set has no more pages to query")
         }
         val (serializedResults, numberOfRowsFromQuery) = resultSetExecutor.execute(serializedParameters, offset)
-        hasNext = limit in 1..numberOfRowsFromQuery
-        offset += limit
+
+        // Here have the number of rows we fetched from the database (before filtering)
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt]: Specify "post-filtering in the database worker" or something along those lines.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt [hunk 17]
```diff
@@ -32,8 +42,20 @@ data class ResultSetImpl<R> internal constructor(
...
+        // then we check if the remainder is 0 when diving the number of rows fetched with the limit
+        // This will:
+        // 1. yield false only if we don't have a next page to fetch
+        // Examples:
+        // `limit` = 5, `numberOfRowsFromQuery` = 9, `hasNext = 9 % 5 == 0` => `hasNext = 4 == 0` == false
+        // `limit` = 5, `numberOfRowsFromQuery` = 4, `hasNext = 4 % 5 == 0` => `hasNext = 4 == 0` == false
+        // 2. yield true if we can't know for sure if there's a next page
+        // Example:
+        // `limit` = 5, `numberOfRowsFromQuery` = 5, `hasNext = 5 % 5 == 0` => `hasNext =  0 == 0` == true
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt]: Show the 10 example as it shows we've covered and understood that scenario.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt [hunk 18]
```diff
@@ -32,8 +42,20 @@ data class ResultSetImpl<R> internal constructor(
...
+        hasNext = numberOfRowsFromQuery != 0  // If there are no rows left there are no more pages to fetch
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/ResultSetImpl.kt:46]: I think put the condition on one line, I don't think we need the comment since the condition is pretty clear now.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 19]
```diff
@@ -72,11 +73,78 @@ class VaultNamedQueryExecutorImpl(
             deserializedParams
         )?.results?.filterNotNull() ?: filteredAndTransformedResults
 
-        // Return the filtered/transformed/collected (if present) result to the caller
+        // Return the filtered/transformed/collected (if present) result and the offset to continue the paging from to the caller
         return EntityResponse.newBuilder()
             .setResults(collectedResults.map { ByteBuffer.wrap(serializationService.serialize(it).bytes) })
-            .setMetadata(KeyValuePairList(listOf(KeyValuePair("numberOfRowsFromQuery", contractStateResults.size.toString()))))
-            .build()
+            .setMetadata(KeyValuePairList(listOf(
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:79]: Can we calculate the `numberOfRowsFromQuery` into a variable then pass to here. Hopefully we can then keep the formatting of this line tidy.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 20]
```diff
@@ -72,11 +73,78 @@ class VaultNamedQueryExecutorImpl(
...
+
+            // Apply filters (if there's any) and add the filtered results to the final result set
+            filteredResults.addAll(contractStateResults.filter {
+                vaultNamedQuery.filter.filter(it, deserializedParams)
+            })
+        }
+
+        return Pair(
+            filteredResults.take(request.limit), // Make sure we don't return more than the limit even if the list has overflown
+            internalOffset - (filteredResults.size - request.limit) // subtract the extra size we removed from the results from the offset
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: We could do `.coerseAtLeast(request.offset)` if we're worried about making any mistakes. This also means the other early return could be removed (but I'm fine leaving it in).

We don't have to do it, I'll let you make the decision.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 21]
```diff
@@ -212,4 +221,9 @@ class VaultNamedQueryExecutorImpl(
             throw NullParameterException(msg)
         }
     }
+
+    data class FilterResult(
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: make this class private.

**Review by lankydan — COMMENTED** [review]
One small comment but otherwise looks good to me.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt (hunk 21 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt]: good point, modified

**Review by lankydan — APPROVED** [review]

**Review by LWogan — APPROVED** [review]
LGTM

**Comment by nkovacsx** [comment]
We discussed the new logic and @lankydan 's outstanding comments on a huddle a week or so ago. Marking them as addressed now.
